File: sub-projects/intelitor-v5.2/lib/indrajaal/holon/database_path.ex
defmodule Indrajaal.Holon.DatabasePath do
  @moduledoc """
  Universal Holon Database Path Resolution.

  ## WHAT
  Provides deterministic path resolution for all holon-specific databases
  using the Universal Holon Identifier (UHI) naming system.

  ## WHY
  - Ensures consistent database path structure across all holons
  - Enables cross-holon database discovery
  - Supports migration from legacy paths
  - Enforces SC-DBNAME-001 to SC-DBNAME-010 constraints

  ## CONSTRAINTS
  - SC-DBNAME-001: All holon databases MUST follow UHI naming
  - SC-DBNAME-002: FQDN resolution MUST be deterministic
  - SC-DBNAME-008: Cross-runtime access MUST use Zenoh
  - SC-DBNAME-009: LOCAL access MUST be direct (no Zenoh)

  ## UHI Format
  {runtime}:{layer}:{domain}:{type}:{instance}

  Examples:
    "ex:l3:kms:srv:main"           -> Elixir L3 KMS Service
    "fs:l4:prj:agt:cockpit"        -> F# L4 Prajna Agent
    "ex:l5:grd:reg:guardian"       -> Elixir L5 Guardian Register

  ## FQDN Format
  {UHI}:{database_type}

  Examples:
    "ex:l3:kms:srv:main:state"     -> state.sqlite
    "ex:l3:kms:srv:main:history"   -> history.duckdb
  """

  @base_path "data/holons"

  # Runtime codes
  @runtimes ~w(ex fs zig rs)

  # Fractal layers
  @layers ~w(l0 l1 l2 l3 l4 l5 l6 l7)

  # Domain codes (registered domains)
  @domains %{
    "kms" => "Knowledge Management System",
    "prj" => "Prajna C3I Cockpit",
    "grd" => "Guardian Safety Kernel",
    "snt" => "Sentinel Health Monitor",
    "imm" => "Immutable Register",
    "fnd" => "Founder Directive",
    "zen" => "Zenoh Communication",
    "bio" => "Biomorphic Systems",
    "pln" => "Planning System",
    "evo" => "Evolution Engine",
    "ctx" => "Cortex AI",
    "tst" => "Test Infrastructure",
    "dev" => "Developer Knowledge",
    "sre" => "SRE Operations",
    "prd" => "Product Lifecycle",
    "obs" => "Observability"
  }

  # Holon types
  @types ~w(srv agt reg str brg pub sub wrk)

  # Database file types
  @db_types %{
    "state" => "state.sqlite",
    "history" => "history.duckdb",
    "vectors" => "vectors.sqlite",
    "cache" => "cache.sqlite",
    "register" => "register.duckdb",
    "analytics" => "analytics.duckdb"
  }

  # ============================================================================
  # Public API
  # ============================================================================

  @doc """
  Resolves a Fully Qualified Database Name (FQDN) to a file path.

  ## Examples

      iex> DatabasePath.resolve("ex:l3:kms:srv:main:state")
      {:ok, "data/holons/ex/l3/kms/main/state.sqlite"}

      iex> DatabasePath.resolve("invalid")
      {:error, :invalid_fqdn}
  """
  @spec resolve(String.t()) :: {:ok, String.t()} | {:error, atom()}
  def resolve(fqdn) when is_binary(fqdn) do
    case parse_fqdn(fqdn) do
      {:ok,
       %{runtime: runtime, layer: layer, domain: domain, instance: instance, db_type: db_type}} ->
        file_name = Map.get(@db_types, db_type)
        path = Path.join([@base_path, runtime, layer, domain, instance, file_name])
        {:ok, path}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Resolves FQDN to path, raising on error.
  """
  @spec resolve!(String.t()) :: String.t()
  def resolve!(fqdn) do
    case resolve(fqdn) do
      {:ok, path} -> path
      {:error, reason} -> raise ArgumentError, "Invalid FQDN #{fqdn}: #{reason}"
    end
  end

  @doc """
  Builds a UHI from components.

  ## Examples

      iex> DatabasePath.build_uhi(:elixir, :l3, :kms, :srv, "main")
      {:ok, "ex:l3:kms:srv:main"}
  """
  @spec build_uhi(atom(), atom(), atom(), atom(), String.t()) ::
          {:ok, String.t()} | {:error, atom()}
  def build_uhi(runtime, layer, domain, type, instance) do
    runtime_code = runtime_to_code(runtime)
    layer_code = Atom.to_string(layer)
    domain_code = Atom.to_string(domain)
    type_code = Atom.to_string(type)

    with :ok <- validate_runtime(runtime_code),
         :ok <- validate_layer(layer_code),
         :ok <- validate_domain(domain_code),
         :ok <- validate_type(type_code),
         :ok <- validate_instance(instance) do
      {:ok, "#{runtime_code}:#{layer_code}:#{domain_code}:#{type_code}:#{instance}"}
    end
  end

  @doc """
  Builds a FQDN from UHI and database type.
  """
  @spec build_fqdn(String.t(), atom()) :: {:ok, String.t()} | {:error, atom()}
  def build_fqdn(uhi, db_type) when is_atom(db_type) do
    db_type_str = Atom.to_string(db_type)

    if Map.has_key?(@db_types, db_type_str) do
      {:ok, "#{uhi}:#{db_type_str}"}
    else
      {:error, :invalid_db_type}
    end
  end

  @doc """
  Returns the directory path for a holon (without database file).

  ## Examples

      iex> DatabasePath.holon_dir("ex:l3:kms:srv:main")
      {:ok, "data/holons/ex/l3/kms/main"}
  """
  @spec holon_dir(String.t()) :: {:ok, String.t()} | {:error, atom()}
  def holon_dir(uhi) when is_binary(uhi) do
    case parse_uhi(uhi) do
      {:ok, %{runtime: runtime, layer: layer, domain: domain, instance: instance}} ->
        path = Path.join([@base_path, runtime, layer, domain, instance])
        {:ok, path}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Returns all database paths for a holon.

  ## Examples

      iex> DatabasePath.all_databases("ex:l3:kms:srv:main")
      {:ok, %{
        state: "data/holons/ex/l3/kms/main/state.sqlite",
        history: "data/holons/ex/l3/kms/main/history.duckdb",
        vectors: "data/holons/ex/l3/kms/main/vectors.sqlite"
      }}
  """
  @spec all_databases(String.t()) :: {:ok, map()} | {:error, atom()}
  def all_databases(uhi) when is_binary(uhi) do
    case holon_dir(uhi) do
      {:ok, dir} ->
        databases =
          @db_types
          |> Enum.map(fn {type, file} ->
            {String.to_atom(type), Path.join(dir, file)}
          end)
          |> Map.new()

        {:ok, databases}

      {:error, reason} ->
        {:error, reason}
    end
  end

  @doc """
  Checks if a UHI represents an Elixir holon.
  """
  @spec elixir_holon?(String.t()) :: boolean()
  def elixir_holon?(uhi) when is_binary(uhi) do
    String.starts_with?(uhi, "ex:")
  end

  @doc """
  Checks if a UHI represents an F# holon.
  """
  @spec fsharp_holon?(String.t()) :: boolean()
  def fsharp_holon?(uhi) when is_binary(uhi) do
    String.starts_with?(uhi, "fs:")
  end

  @doc """
  Returns the Zenoh topic for cross-holon database access.

  ## Examples

      iex> DatabasePath.zenoh_topic("ex:l3:kms:srv:main", :query)
      "indrajaal/db/ex/l3/kms/main/query"
  """
  @spec zenoh_topic(String.t(), atom()) :: String.t()
  def zenoh_topic(uhi, operation) when is_binary(uhi) and is_atom(operation) do
    {:ok, %{runtime: runtime, layer: layer, domain: domain, instance: instance}} = parse_uhi(uhi)
    "indrajaal/db/#{runtime}/#{layer}/#{domain}/#{instance}/#{operation}"
  end

  @doc """
  Maps a legacy path to the new naming system.
  """
  @spec from_legacy(String.t()) :: {:ok, String.t()} | {:error, :unknown_legacy_path}
  def from_legacy(legacy_path) do
    legacy_mappings = %{
      "data/kms/holons.db" => "ex:l3:kms:srv:main:state",
      "data/kms/analytics.duckdb" => "ex:l3:kms:srv:main:history",
      "data/holons/founder_directive/state.sqlite" => "ex:l5:fnd:reg:founder:state",
      "data/holons/founder_directive/history.duckdb" => "ex:l5:fnd:reg:founder:history",
      "data/holons/prajna_register.duckdb" => "ex:l5:prj:srv:prajna:register",
      "data/smriti/planning.db" => "fs:l4:pln:srv:main:state",
      "data/kms/smriti.db" => "ex:l3:kms:str:smriti:state"
    }

    # Check direct match first
    case Map.get(legacy_mappings, legacy_path) do
      nil ->
        # Check pattern matches for node-specific paths
        cond do
          String.match?(legacy_path, ~r/^data\/kms\/[^\/]+\/holons\.db$/) ->
            node = legacy_path |> String.split("/") |> Enum.at(2)
            {:ok, "ex:l3:kms:srv:#{node}:state"}

          String.match?(legacy_path, ~r/^data\/kms\/[^\/]+\/analytics\.duckdb$/) ->
            node = legacy_path |> String.split("/") |> Enum.at(2)
            {:ok, "ex:l3:kms:srv:#{node}:history"}

          String.match?(legacy_path, ~r/^data\/smriti\/[^\/]+\/holons\.db$/) ->
            node = legacy_path |> String.split("/") |> Enum.at(2)
            {:ok, "ex:l3:kms:srv:#{node}:state"}

          true ->
            {:error, :unknown_legacy_path}
        end

      fqdn ->
        {:ok, fqdn}
    end
  end

  @doc """
  Returns all registered domain codes.
  """
  @spec domains() :: map()
  def domains, do: @domains

  @doc """
  Returns all valid database types.
  """
  @spec db_types() :: map()
  def db_types, do: @db_types

  # ============================================================================
  # Private Functions
  # ============================================================================

  defp parse_fqdn(fqdn) do
    parts = String.split(fqdn, ":")

    case parts do
      [runtime, layer, domain, type, instance, db_type] ->
        with :ok <- validate_runtime(runtime),
             :ok <- validate_layer(layer),
             :ok <- validate_domain(domain),
             :ok <- validate_type(type),
             :ok <- validate_instance(instance),
             :ok <- validate_db_type(db_type) do
          {:ok,
           %{
             runtime: runtime,
             layer: layer,
             domain: domain,
             type: type,
             instance: instance,
             db_type: db_type
           }}
        end

      _ ->
        {:error, :invalid_fqdn_format}
    end
  end

  defp parse_uhi(uhi) do
    parts = String.split(uhi, ":")

    case parts do
      [runtime, layer, domain, type, instance] ->
        with :ok <- validate_runtime(runtime),
             :ok <- validate_layer(layer),
             :ok <- validate_domain(domain),
             :ok <- validate_type(type),
             :ok <- validate_instance(instance) do
          {:ok,
           %{
             runtime: runtime,
             layer: layer,
             domain: domain,
             type: type,
             instance: instance
           }}
        end

      _ ->
        {:error, :invalid_uhi_format}
    end
  end

  defp runtime_to_code(:elixir), do: "ex"
  defp runtime_to_code(:fsharp), do: "fs"
  defp runtime_to_code(:zig), do: "zig"
  defp runtime_to_code(:rust), do: "rs"
  defp runtime_to_code(atom) when is_atom(atom), do: Atom.to_string(atom)

  defp validate_runtime(runtime) when runtime in @runtimes, do: :ok
  defp validate_runtime(_), do: {:error, :invalid_runtime}

  defp validate_layer(layer) when layer in @layers, do: :ok
  defp validate_layer(_), do: {:error, :invalid_layer}

  defp validate_domain(domain) do
    if Map.has_key?(@domains, domain), do: :ok, else: {:error, :invalid_domain}
  end

  defp validate_type(type) when type in @types, do: :ok
  defp validate_type(_), do: {:error, :invalid_type}

  defp validate_instance(instance) when is_binary(instance) and byte_size(instance) > 0, do: :ok
  defp validate_instance(_), do: {:error, :invalid_instance}

  defp validate_db_type(db_type) do
    if Map.has_key?(@db_types, db_type), do: :ok, else: {:error, :invalid_db_type}
  end
end

File: intelitor-v5.2/lib/cepaf/src/Cepaf.Holon/DatabasePath.fs
/// Universal Holon Database Path Resolution for F#
///
/// WHAT: Provides deterministic path resolution for all holon-specific databases
///       using the Universal Holon Identifier (UHI) naming system.
///
/// WHY:
/// - Ensures consistent database path structure across all holons
/// - Enables cross-holon database discovery
/// - Supports migration from legacy paths
/// - Enforces SC-DBNAME-001 to SC-DBNAME-010 constraints
///
/// CONSTRAINTS:
/// - SC-DBNAME-001: All holon databases MUST follow UHI naming
/// - SC-DBNAME-002: FQDN resolution MUST be deterministic
/// - SC-DBNAME-008: Cross-runtime access MUST use Zenoh
/// - SC-DBNAME-009: LOCAL access MUST be direct (no Zenoh)
///
/// UHI Format: {runtime}:{layer}:{domain}:{type}:{instance}
///
/// Examples:
///   "ex:l3:kms:srv:main"           -> Elixir L3 KMS Service
///   "fs:l4:prj:agt:cockpit"        -> F# L4 Prajna Agent
///   "ex:l5:grd:reg:guardian"       -> Elixir L5 Guardian Register
module Cepaf.Holon.DatabasePath

open System
open System.IO

// ============================================================================
// Types
// ============================================================================

/// Runtime type enumeration
type Runtime =
    | Elixir
    | FSharp
    | Zig
    | Rust

/// Fractal layer enumeration
type FractalLayer =
    | L0  // Runtime
    | L1  // Function
    | L2  // Component
    | L3  // Holon
    | L4  // Container
    | L5  // Node
    | L6  // Cluster
    | L7  // Federation

/// Domain codes
type Domain =
    | Kms   // Knowledge Management System
    | Prj   // Prajna C3I Cockpit
    | Grd   // Guardian Safety Kernel
    | Snt   // Sentinel Health Monitor
    | Imm   // Immutable Register
    | Fnd   // Founder Directive
    | Zen   // Zenoh Communication
    | Bio   // Biomorphic Systems
    | Pln   // Planning System
    | Evo   // Evolution Engine
    | Ctx   // Cortex AI
    | Tst   // Test Infrastructure
    | Dev   // Developer Knowledge
    | Sre   // SRE Operations
    | Prd   // Product Lifecycle
    | Obs   // Observability

/// Holon type
type HolonType =
    | Srv   // Service
    | Agt   // Agent
    | Reg   // Register/Registry
    | Str   // Store
    | Brg   // Bridge
    | Pub   // Publisher
    | Sub   // Subscriber
    | Wrk   // Worker

/// Database type
type DatabaseType =
    | State     // state.sqlite
    | History   // history.duckdb
    | Vectors   // vectors.sqlite
    | Register  // register.duckdb
    | Analytics // analytics.duckdb

/// Universal Holon Identifier
type UHI = {
    Runtime: Runtime
    Layer: FractalLayer
    Domain: Domain
    Type: HolonType
    Instance: string
}

/// Fully Qualified Database Name
type FQDN = {
    UHI: UHI
    DatabaseType: DatabaseType
}

/// Parse result
type ParseResult<'T> =
    | Success of 'T
    | Error of string

// ============================================================================
// Constants
// ============================================================================

let private basePath = "data/holons"

let private dbTypeToFile = function
    | State -> "state.sqlite"
    | History -> "history.duckdb"
    | Vectors -> "vectors.sqlite"
    | Register -> "register.duckdb"
    | Analytics -> "analytics.duckdb"

let private runtimeToCode = function
    | Elixir -> "ex"
    | FSharp -> "fs"
    | Zig -> "zig"
    | Rust -> "rs"

let private codeToRuntime = function
    | "ex" -> Some Elixir
    | "fs" -> Some FSharp
    | "zig" -> Some Zig
    | "rs" -> Some Rust
    | _ -> None

let private layerToCode = function
    | L0 -> "l0" | L1 -> "l1" | L2 -> "l2" | L3 -> "l3"
    | L4 -> "l4" | L5 -> "l5" | L6 -> "l6" | L7 -> "l7"

let private codeToLayer = function
    | "l0" -> Some L0 | "l1" -> Some L1 | "l2" -> Some L2 | "l3" -> Some L3
    | "l4" -> Some L4 | "l5" -> Some L5 | "l6" -> Some L6 | "l7" -> Some L7
    | _ -> None

let private domainToCode = function
    | Kms -> "kms" | Prj -> "prj" | Grd -> "grd" | Snt -> "snt"
    | Imm -> "imm" | Fnd -> "fnd" | Zen -> "zen" | Bio -> "bio"
    | Pln -> "pln" | Evo -> "evo" | Ctx -> "ctx" | Tst -> "tst"
    | Dev -> "dev" | Sre -> "sre" | Prd -> "prd" | Obs -> "obs"

let private codeToDomain = function
    | "kms" -> Some Kms | "prj" -> Some Prj | "grd" -> Some Grd | "snt" -> Some Snt
    | "imm" -> Some Imm | "fnd" -> Some Fnd | "zen" -> Some Zen | "bio" -> Some Bio
    | "pln" -> Some Pln | "evo" -> Some Evo | "ctx" -> Some Ctx | "tst" -> Some Tst
    | "dev" -> Some Dev | "sre" -> Some Sre | "prd" -> Some Prd | "obs" -> Some Obs
    | _ -> None

let private typeToCode = function
    | Srv -> "srv" | Agt -> "agt" | Reg -> "reg" | Str -> "str"
    | Brg -> "brg" | Pub -> "pub" | Sub -> "sub" | Wrk -> "wrk"

let private codeToType = function
    | "srv" -> Some Srv | "agt" -> Some Agt | "reg" -> Some Reg | "str" -> Some Str
    | "brg" -> Some Brg | "pub" -> Some Pub | "sub" -> Some Sub | "wrk" -> Some Wrk
    | _ -> None

let private dbTypeToCode = function
    | State -> "state" | History -> "history" | Vectors -> "vectors"
    | Register -> "register" | Analytics -> "analytics"

let private codeToDbType = function
    | "state" -> Some State | "history" -> Some History | "vectors" -> Some Vectors
    | "register" -> Some Register | "analytics" -> Some Analytics
    | _ -> None

// ============================================================================
// UHI Functions
// ============================================================================

/// Creates a UHI from components
let createUHI runtime layer domain htype instance =
    { Runtime = runtime; Layer = layer; Domain = domain; Type = htype; Instance = instance }

/// Converts UHI to string format
let uhiToString (uhi: UHI) =
    sprintf "%s:%s:%s:%s:%s"
        (runtimeToCode uhi.Runtime)
        (layerToCode uhi.Layer)
        (domainToCode uhi.Domain)
        (typeToCode uhi.Type)
        uhi.Instance

/// Parses a UHI string
let parseUHI (s: string) : ParseResult<UHI> =
    let parts = s.Split(':')
    if parts.Length <> 5 then
        Error "Invalid UHI format: expected 5 parts separated by :"
    else
        let runtime = codeToRuntime parts.[0]
        let layer = codeToLayer parts.[1]
        let domain = codeToDomain parts.[2]
        let htype = codeToType parts.[3]
        let instance = parts.[4]

        match runtime, layer, domain, htype with
        | Some r, Some l, Some d, Some t when not (String.IsNullOrEmpty instance) ->
            Success { Runtime = r; Layer = l; Domain = d; Type = t; Instance = instance }
        | None, _, _, _ -> Error $"Invalid runtime code: {parts.[0]}"
        | _, None, _, _ -> Error $"Invalid layer code: {parts.[1]}"
        | _, _, None, _ -> Error $"Invalid domain code: {parts.[2]}"
        | _, _, _, None -> Error $"Invalid type code: {parts.[3]}"
        | _ -> Error "Invalid instance: cannot be empty"

// ============================================================================
// FQDN Functions
// ============================================================================

/// Creates a FQDN from UHI and database type
let createFQDN uhi dbType = { UHI = uhi; DatabaseType = dbType }

/// Converts FQDN to string format
let fqdnToString (fqdn: FQDN) =
    sprintf "%s:%s" (uhiToString fqdn.UHI) (dbTypeToCode fqdn.DatabaseType)

/// Parses a FQDN string
let parseFQDN (s: string) : ParseResult<FQDN> =
    let parts = s.Split(':')
    if parts.Length <> 6 then
        Error "Invalid FQDN format: expected 6 parts separated by :"
    else
        let uhiStr = String.Join(":", parts.[0..4])
        match parseUHI uhiStr with
        | Error e -> Error e
        | Success uhi ->
            match codeToDbType parts.[5] with
            | Some dbType -> Success { UHI = uhi; DatabaseType = dbType }
            | None -> Error $"Invalid database type: {parts.[5]}"

// ============================================================================
// Path Resolution
// ============================================================================

/// Resolves a FQDN to a file path
let resolve (fqdn: FQDN) =
    let uhi = fqdn.UHI
    let fileName = dbTypeToFile fqdn.DatabaseType
    Path.Combine(
        basePath,
        runtimeToCode uhi.Runtime,
        layerToCode uhi.Layer,
        domainToCode uhi.Domain,
        uhi.Instance,
        fileName
    )

/// Resolves a FQDN string to a file path
let resolveString (fqdnStr: string) : ParseResult<string> =
    match parseFQDN fqdnStr with
    | Success fqdn -> Success (resolve fqdn)
    | Error e -> Error e

/// Returns the directory path for a holon
let holonDir (uhi: UHI) =
    Path.Combine(
        basePath,
        runtimeToCode uhi.Runtime,
        layerToCode uhi.Layer,
        domainToCode uhi.Domain,
        uhi.Instance
    )

/// Returns all database paths for a holon
let allDatabases (uhi: UHI) =
    let dir = holonDir uhi
    [
        State, Path.Combine(dir, dbTypeToFile State)
        History, Path.Combine(dir, dbTypeToFile History)
        Vectors, Path.Combine(dir, dbTypeToFile Vectors)
        Register, Path.Combine(dir, dbTypeToFile Register)
        Analytics, Path.Combine(dir, dbTypeToFile Analytics)
    ] |> Map.ofList

// ============================================================================
// Cross-Runtime Helpers
// ============================================================================

/// Checks if UHI is an Elixir holon
let isElixirHolon (uhi: UHI) = uhi.Runtime = Elixir

/// Checks if UHI is an F# holon
let isFSharpHolon (uhi: UHI) = uhi.Runtime = FSharp

/// Returns the Zenoh topic for cross-holon database access
let zenohTopic (uhi: UHI) (operation: string) =
    sprintf "indrajaal/db/%s/%s/%s/%s/%s"
        (runtimeToCode uhi.Runtime)
        (layerToCode uhi.Layer)
        (domainToCode uhi.Domain)
        uhi.Instance
        operation

// ============================================================================
// Legacy Path Migration
// ============================================================================

/// Maps a legacy path to FQDN
let fromLegacy (legacyPath: string) : ParseResult<string> =
    let mappings =
        [ ("data/kms/holons.db", "ex:l3:kms:srv:main:state")
          ("data/kms/analytics.duckdb", "ex:l3:kms:srv:main:history")
          ("data/holons/founder_directive/state.sqlite", "ex:l5:fnd:reg:founder:state")
          ("data/holons/founder_directive/history.duckdb", "ex:l5:fnd:reg:founder:history")
          ("data/holons/prajna_register.duckdb", "ex:l5:prj:srv:prajna:register")
          ("data/smriti/planning.db", "fs:l4:pln:srv:main:state")
          ("data/kms/smriti.db", "ex:l3:kms:str:smriti:state") ]
        |> Map.ofList
    match Map.tryFind legacyPath mappings with
    | Some fqdn -> Success fqdn
    | None ->
        if legacyPath.StartsWith("data/kms/") && legacyPath.EndsWith("/holons.db") then
            let parts = legacyPath.Split('/')
            if parts.Length >= 3 then
                Success $"ex:l3:kms:srv:{parts.[2]}:state"
            else
                Error "Unknown legacy path pattern"
        elif legacyPath.StartsWith("data/kms/") && legacyPath.EndsWith("/analytics.duckdb") then
            let parts = legacyPath.Split('/')
            if parts.Length >= 3 then
                Success $"ex:l3:kms:srv:{parts.[2]}:history"
            else
                Error "Unknown legacy path pattern"
        else
            Error $"Unknown legacy path: {legacyPath}"

// ============================================================================
// Domain Registry
// ============================================================================

/// Returns all registered domains with descriptions
let domainRegistry =
    [ (Kms, "Knowledge Management System")
      (Prj, "Prajna C3I Cockpit")
      (Grd, "Guardian Safety Kernel")
      (Snt, "Sentinel Health Monitor")
      (Imm, "Immutable Register")
      (Fnd, "Founder Directive")
      (Zen, "Zenoh Communication")
      (Bio, "Biomorphic Systems")
      (Pln, "Planning System")
      (Evo, "Evolution Engine")
      (Ctx, "Cortex AI")
      (Tst, "Test Infrastructure")
      (Dev, "Developer Knowledge")
      (Sre, "SRE Operations")
      (Prd, "Product Lifecycle")
      (Obs, "Observability") ]
    |> Map.ofList

// ============================================================================
// Manifest Support
// ============================================================================

/// Manifest structure for holon metadata
type HolonManifest = {
    Version: string
    UHI: string
    FQUN: string
    CreatedAt: DateTime
    UpdatedAt: DateTime
    Runtime: {| Type: string; Version: string |}
    Databases: Map<string, {| Type: string; Version: string; SchemaVersion: int |}>
    Capabilities: string list
    ParentUHI: string option
    ChildrenUHI: string list
    ZenohTopics: {| Publish: string list; Subscribe: string list |}
    Checksum: string
}

/// Creates a default manifest for a holon
let createManifest (uhi: UHI) =
    let uhiStr = uhiToString uhi
    let now = DateTime.UtcNow
    {
        Version = "1.0.0"
        UHI = uhiStr
        FQUN = $"kms/l3/{domainToCode uhi.Domain}/default/{uhi.Instance}"
        CreatedAt = now
        UpdatedAt = now
        Runtime = {| Type = runtimeToCode uhi.Runtime; Version = "10.0.0" |}
        Databases = Map.ofList [
            "state", {| Type = "sqlite"; Version = "3.47.0"; SchemaVersion = 1 |}
            "history", {| Type = "duckdb"; Version = "1.2.0"; SchemaVersion = 1 |}
        ]
        Capabilities = ["read"; "write"]
        ParentUHI = None
        ChildrenUHI = []
        ZenohTopics = {|
            Publish = [zenohTopic uhi "state"]
            Subscribe = ["indrajaal/coord/heartbeat"]
        |}
        Checksum = ""
    }
